incubator/hnc/pkg/controllers/hierarchy_controller.go

```diff
@@ -76,7 +88,11 @@ type NamespaceSyncer interface {
 func (r *HierarchyReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background()
 	ns := req.NamespacedName.Namespace
-	log := r.Log.WithValues("ns", ns)
+
+	id := r.lockNamespace(ns)
```

Comment:
I don't understand why reconcile can be called by multiple threads, can you elaborate it ?

Reply:
All controllers are multithreaded ([source](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L168)). When the HNC calls enqueueAffected on other namespaces, those other namespaces can start being reconciled immediately. I've observed this happening in the log files; different reconciliation calls definitely work in parallel.

Reply:
I see, we usually use default MaxConcurrentReconciles = 1 to avoid multi-threading problem. Otherwise, you probably need lock for every reconcile object like what you did in this change for thread safe. Just curious what parallel level you are going to specify in HNC?

Reply:
No idea yet but I'd imagine that with any significantly sized cluster, we'll need more than one thread.

Reply:
Hmm, looks like the default is 1, I'm not sure why I thought I was seeing multiple concurrent threads. I'll investigate more closely in a subsequent PR.